### ISSUE TITLE
High CPU Usage and memory leak

### ISSUE THREAD

**Issue by gustavomassa** [dashersw/cote#118]
Hello,

I'm facing an issue about high CPU Usage (about 30%) and a memory leak over time.
The app is using node appmetrics to get real time process metrics and sending the data over network(cote) every 5 seconds to a specific monitoring service.

![image](https://user-images.githubusercontent.com/16235598/45715892-af323c80-bb6b-11e8-8448-8f4923bef96c.png)
![image](https://user-images.githubusercontent.com/16235598/45715921-c6712a00-bb6b-11e8-91b0-adbe113e47e7.png)

![image](https://user-images.githubusercontent.com/16235598/45715772-667a8380-bb6b-11e8-836d-dc5ca573bc5e.png)
![image](https://user-images.githubusercontent.com/16235598/45715795-7b571700-bb6b-11e8-9645-40890f76bf6e.png)

These are results of the app running on Linux (CentOS 7), the results running on windows are worse.

**Comment by dashersw** [comment]
Hello @gustavomassa , do you have a code example that you could share to replicate the issue?

**Comment by gustavomassa** [comment]
Hey @dashersw sure!

I've found one of or the main root cause of the high cpu usage.
Before going into the code let me explain the user case and the environment settings:

First of all, here we use an interval environment config that being used for COTE_ENV as well, that environment is actually the build type: dev, qa, prod.
The second filter for the microservices communication is the namespace, the namespace is being set as the name of the current service.
The third filter is the key property, that always has the hostname of the service, so actually our architecture can transfer services messages just from the same machine.

Example:

**Microservice 1**
- environment: "dev"
- namespace: "WBA-MAILER"
- key: "ITJ001"

**Microservice 2**
- environment: "dev"
- namespace: "WBA-MONITOR"
- key: "ITJ002"

User case: 
WBA-MONITOR receives ping responses when requests service discovery broadcast(ping request and wait 5 seconds for all received pong and push a list of online services)
WBA-MONITOR receives telemetry data every 5 seconds, "loop" and "eventloop" data objects.

WBA-MAILER responds a ping request with pong
WBA-MAILER publish telemetry data based on appmetrics every 5 seconds to "loop" and "eventloop" topics.

Each microservice has its own service controller:
```
import * as uuid from 'uuid';
import { Readable } from 'stream';
import { WBA } from './Webbuffet';
import { ServiceRepository } from './persistence/ServiceRepository';
import { DiscoveryOptions, Requester, Responder, Publisher, Subscriber } from 'cote';

export interface ServiceEvent {
    messageId: string;
    from: {
        hostname: string;
        name: string;
        environment: string;
        version: string;
        coreVersion: string;
    };
    to: string;
    type: string;
    date: Date,
    data: any;
    timeout?: number;
}

export interface ServiceLogEvent extends ServiceEvent {
    //event: string, //request, response, publish, subescribe
    receivedBy: {
        hostname: string;
        name: string;
        environment: string;
        version: string;
        coreVersion: string;
    };
    receivedDate: Date
}

export interface EmailMessage {
    from: WBA.Email | string;
    to: string;
    subject: string;
    html: string;
    attachments?: Array<EmailAttachment>;
}

export interface EmailAttachment {
    filename: string;
    content: string | Buffer | Readable;
    encoding?: string;
    contentType?: string;
}


export class ServiceController {
    private static _instance: ServiceController;
    private cote: any;
    private discoveryOptions: DiscoveryOptions;
    private res: Responder;
    private sub: Subscriber;
    private broadcastSender: Publisher;
    private broadcastReceiver: Subscriber;
    private telemetrySender: Publisher;


    private constructor() {
        const config = WBA.config();
        this.cote = require('cote')({ environment: config.environment, broadcast: '192.168.3.255' });
        //this.cote = require('cote')({ environment: config.environment, multicast: '230.0.0.0' });

        /*          default values
                    helloInterval: 2000,
                    checkInterval: 4000,
                    nodeTimeout: 5000,
                    masterTimeout: 6000,
                    monitor: false,
                    log: true,
                    helloLogsEnabled: true,
                    statusLogsEnabled: true,
                    ignoreProcess: false */

        this.discoveryOptions = {
            ignoreProcess: true,
            monitor: false,
            log: false

        };
        this.res = new this.cote.Responder({ name: `${config.name} Responder`, namespace: config.name, key: config.hostname }, this.discoveryOptions);
        this.sub = new this.cote.Subscriber({ name: `${config.name} Subscriber`, namespace: config.name, key: config.hostname }, this.discoveryOptions);
        this.broadcastSender = new this.cote.Publisher({ name: `${config.name} Broadcast`, namespace: config.hostname }, this.discoveryOptions);
        this.broadcastReceiver = new this.cote.Subscriber({ name: `${config.name} Broadcast Receiver`, namespace: config.hostname }, this.discoveryOptions);
        this.telemetrySender = new this.cote.Publisher({ name: `${WBA.config().name} Monitor Publisher`, namespace: 'WBA-MONITOR', key: WBA.config().hostname }, this.discoveryOptions);

        const rc: WBA.Webbuffet = this.registerServiceDiscovery();
        if (rc.$status != WBA.Status.SUCCESS) WBA.logError(new Error(`Failed to register ping response: ${JSON.stringify(rc)}`));
    }

    //singleton get instance
    public static get Instance(): ServiceController {
        return this._instance || (this._instance = new this());
    }

    public async init(): Promise<void> {
        try {


        } catch (ex) {
            WBA.logError(new Error(ex));
        }
    }

    private pingResponse(msg: ServiceEvent): void {
        try {
            //publish to pong
            //const rc = this.publishMsg(msg.from.name, 'pong', 'pong');
            const rc = this.sendTelemetry('pong', 'pong');
            if (rc.$status != WBA.Status.SUCCESS) WBA.logError(new Error(`Failed to publish ping response: ${JSON.stringify(rc)}`));

        } catch (ex) {
            WBA.logError(new Error(ex));
        }
    }

    private registerServiceDiscovery(): WBA.Webbuffet {
        try {
            WBA.log(`Registered Ping Broadcast Response`);

            this.broadcastReceiver.on('ping', (msg: ServiceEvent) => this.pingResponse(msg));

            return WBA.success(true);

        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }

    private async persistReceivedMessage(request: ServiceEvent): Promise<void> {
        try {
            //ignore metricts and service discovery events
            if (request.type === 'ping' || request.type === 'pong' || request.type === 'loop' || request.type === 'eventloop' || request.type === 'cpu' || request.type === 'memory' || request.type === 'profiling') return;

            //persist message
            const log: ServiceLogEvent = {
                messageId: request.messageId,
                from: {
                    environment: request.from.environment,
                    hostname: request.from.hostname,
                    name: request.from.name,
                    version: request.from.version,
                    coreVersion: request.from.coreVersion
                },
                to: request.to,
                type: request.type,
                date: new Date(request.date),
                data: request.data,
                timeout: request.timeout,
                receivedBy: {
                    environment: WBA.config().environment,
                    hostname: WBA.config().hostname,
                    name: WBA.config().name,
                    version: WBA.config().version,
                    coreVersion: WBA.getCoreVersion()
                },
                receivedDate: new Date()
            }
            const repository = new ServiceRepository();
            const rc = await repository.insertLog(log);
            if (rc.$status != WBA.Status.SUCCESS) WBA.logError(new Error(`Failed to persist received message: ${JSON.stringify(rc)}`));

        } catch (ex) {
            WBA.logError(new Error(ex));
        }
    }

    public sendTelemetry(type: string, msg: any): WBA.Webbuffet {
        try {
            if (!type) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'type is NULL');

            const payload: ServiceEvent = {
                messageId: uuid.v1(),
                from: {
                    environment: WBA.config().environment,
                    hostname: WBA.config().hostname,
                    name: WBA.config().name,
                    version: WBA.config().version,
                    coreVersion: WBA.getCoreVersion()
                },
                to: 'WBA-MONITOR',
                type: type,
                date: new Date(),
                data: msg
            };
            this.telemetrySender.publish(type, payload);

            return WBA.success(true);
        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }

    public publishMsg(namespace: string, topic: string, msg: any): WBA.Webbuffet {
        try {
            //if (!namespace || !topic || !msg) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'namespace/topic/msg is NULL');
            if (!namespace || !topic) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'namespace/topic is NULL');

            const publisher: Publisher = new this.cote.Publisher({ name: `${WBA.config().name} Publisher`, namespace: namespace, key: WBA.config().hostname }, this.discoveryOptions);
            const payload: ServiceEvent = {
                messageId: uuid.v1(),
                from: {
                    environment: WBA.config().environment,
                    hostname: WBA.config().hostname,
                    name: WBA.config().name,
                    version: WBA.config().version,
                    coreVersion: WBA.getCoreVersion()
                },
                to: namespace,
                type: topic,
                date: new Date(),
                data: msg
            };
            publisher.publish(topic, payload);

            return WBA.success(true);
        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }

    public serviceDiscoveryRequest(): WBA.Webbuffet {
        try {
            return this.broadcastMsg('ping', null);

        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }

    public broadcastMsg(topic: string, msg: any): WBA.Webbuffet {
        try {
            //if (!topic || !msg) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'topic/msg is NULL');
            if (!topic) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'topic is NULL');

            //const broadcaster: Publisher = new this.cote.Publisher({ name: `${WBA.config().name} Broadcast`, namespace: WBA.config().hostname }, this.discoveryOptions);
            const payload: ServiceEvent = {
                messageId: uuid.v1(),
                from: {
                    environment: WBA.config().environment,
                    hostname: WBA.config().hostname,
                    name: WBA.config().name,
                    version: WBA.config().version,
                    coreVersion: WBA.getCoreVersion()
                },
                to: 'broadcast',
                type: topic,
                date: new Date(),
                data: msg
            };
            this.broadcastSender.publish(topic, payload);

            return WBA.success(true);
        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }

    public async sendMsg(namespace: string, topic: string, msg: any, timeout?: number): Promise<WBA.Webbuffet> {
        try {
            //if (!namespace || !topic || !msg) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'namespace/topic/msg is NULL');
            if (!namespace || !topic) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'namespace/topic is NULL');

            const requester: Requester = new this.cote.Requester({ name: `${WBA.config().name} Requester`, namespace: namespace, key: WBA.config().hostname }, this.discoveryOptions);
            const operationTimeout = (timeout) ? timeout : 15000;
            const payload: ServiceEvent = {
                messageId: uuid.v1(),
                from: {
                    environment: WBA.config().environment,
                    hostname: WBA.config().hostname,
                    name: WBA.config().name,
                    version: WBA.config().version,
                    coreVersion: WBA.getCoreVersion()
                },
                to: namespace,
                type: topic,
                date: new Date(),
                data: msg,
                timeout: operationTimeout
            };

            //send request
            const response: Object = await requester.send(payload).timeout(operationTimeout);
            if (response && response.hasOwnProperty('status') && response.hasOwnProperty('data')) {
                let wbaResponse = WBA.factory();
                wbaResponse.$status = response['status'];
                wbaResponse.$data = response['data'];
                wbaResponse.$hasError = response['hasError'];
                return wbaResponse;
            } else return WBA.error(WBA.Status.FAILED, `Null/Invalid response from service: ${namespace} on topic: ${topic}`);

        } catch (ex) {
            if (ex.name && ex.name === 'TimeoutError') return WBA.error(WBA.Status.PROMISE_TIMEOUT, new Error(ex).message);
            else if (ex.hasOwnProperty('status') && ex.hasOwnProperty('data')) {//wba type
                let wbaResponse = WBA.factory();
                wbaResponse.$status = ex['status'];
                wbaResponse.$data = ex['data'];
                wbaResponse.$hasError = ex['hasError'];
                return wbaResponse;
            } else return WBA.exception(new Error(ex));
        }
    }

    public registerResponder(topic: string, responseHandler: (request: ServiceEvent) => Promise<WBA.Webbuffet>): WBA.Webbuffet {
        try {
            if (!topic || !responseHandler) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'topic/responseHandler is NULL');

            WBA.log(`Registered responder: ${topic}`);

            this.res.on(topic, async (request: ServiceEvent) => {
                try {
                    //ignore received messages from different hostnames when in dev envinronment
                    //if (WBA.config().environment === 'dev' && request.from.hostname !== WBA.config().hostname) return;

                    this.persistReceivedMessage(request);

                    const timeout = (request.timeout) ? request.timeout : 15000;
                    const response = await responseHandler(request).timeout(timeout);
                    return response;

                } catch (ex) {
                    if (ex.name && ex.name === 'TimeoutError') return WBA.error(WBA.Status.PROMISE_TIMEOUT, new Error(ex).message);
                    else if (ex.hasOwnProperty('status') && ex.hasOwnProperty('data')) {//wba type
                        let wbaResponse = WBA.factory();
                        wbaResponse.$status = ex['status'];
                        wbaResponse.$data = ex['data'];
                        wbaResponse.$hasError = ex['hasError'];
                        return wbaResponse;
                    } else return WBA.exception(new Error(ex));
                }
            });

            return WBA.success(true);

        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }

    public registerSubscriber(topic: string, msgHandler: (msg: ServiceEvent) => void | Promise<void>): WBA.Webbuffet {
        try {
            if (!topic || !msgHandler) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'topic/msgHandler is NULL');

            WBA.log(`Registered subscriber: ${topic}`);

            this.sub.on(topic, (msg: ServiceEvent) => {
                //ignore received messages from different hostnames when in dev envinronment
                //if (WBA.config().environment === 'dev' && msg.from.hostname !== WBA.config().hostname) return;

                this.persistReceivedMessage(msg);
                msgHandler(msg);
            });

            return WBA.success(true);

        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }

    public registerBroadcast(topic: string, msgHandler: (msg: ServiceEvent) => void | Promise<void>): WBA.Webbuffet {
        try {
            if (!topic || !msgHandler) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'topic/msgHandler is NULL');

            WBA.log(`Registered Ping Broadcast Response: ${topic}`);

            this.broadcastReceiver.on(topic, (msg: ServiceEvent) => {
                //ignore received messages from different hostnames when in dev envinronment
                //if (WBA.config().environment === 'dev' && msg.from.hostname !== WBA.config().hostname) return;

                this.persistReceivedMessage(msg);
                msgHandler(msg);
            });

            return WBA.success(true);

        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }


    public async requestEmail(email: EmailMessage): Promise<WBA.Webbuffet> {
        try {
            if (!email) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'email is NULL');
            if (!email.from || !email.to || !email.subject || !email.html) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'from/to/subject/body is NULL');


            let sender = null;
            if (typeof (email.from) === 'string') sender = email.from;
            else {
                switch (<WBA.Email>email.from) {
                    case WBA.Email.TINA:
                        sender = 'TINA Here <tina@allog.com.br>';
                        break;
                    case WBA.Email.ALLOG:
                        sender = 'TINA Here <tina@allog.com.br>';
                        break;
                    case WBA.Email.MONACO:
                        sender = 'MONACO <wba@allog.com.br>';
                        break;
                }
            }
            email.from = sender;
            return this.sendMsg(WBA.config().mailer_service, 'emailRequest', email, 30000);

        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }
}
```
Part of the metric controller responsible for sending the appmetrics telemetry data every 5 seconds:

```
this.monitoring.on('eventloop', (eventLoop) => {
            //serviceController.publishMsg('WBA-MONITOR', 'eventloop', eventLoop);
            serviceController.sendTelemetry('eventloop', eventLoop);
        });

        this.monitoring.on('loop', (loop) => {
            //serviceController.publishMsg('WBA-MONITOR', 'loop', loop);
            serviceController.sendTelemetry('loop', loop);
        });
```

That code is working fine, the commented function, publishMsg was generating high cpu usage, let me explain:
Every 5 seconds the metric controller send 2 objects to service controller(cote), inside cote it was calling that function:
    ```
public publishMsg(namespace: string, topic: string, msg: any): WBA.Webbuffet {
        try {
            //if (!namespace || !topic || !msg) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'namespace/topic/msg is NULL');
            if (!namespace || !topic) return WBA.error(WBA.Status.ERROR_NULL_PARAMETER, 'namespace/topic is NULL');

            const publisher: Publisher = new this.cote.Publisher({ name: `${WBA.config().name} Publisher`, namespace: namespace, key: WBA.config().hostname }, this.discoveryOptions);
            const payload: ServiceEvent = {
                messageId: uuid.v1(),
                from: {
                    environment: WBA.config().environment,
                    hostname: WBA.config().hostname,
                    name: WBA.config().name,
                    version: WBA.config().version,
                    coreVersion: WBA.getCoreVersion()
                },
                to: namespace,
                type: topic,
                date: new Date(),
                data: msg
            };
            publisher.publish(topic, payload);

            return WBA.success(true);
        } catch (ex) {
            return WBA.exception(new Error(ex));
        }
    }
```
So basically the cpu usage issue is regarding instantiating a new publisher object dynamically every 5 seconds to send the telemetry data:
`const publisher: Publisher = new this.cote.Publisher({ name: `${WBA.config().name} Publisher`, namespace: namespace, key: WBA.config().hostname }, this.discoveryOptions);`
I fixed that creating the telemetry publisher that always use the WBA-MONITOR namespace, so it just instantiate one telemetry publisher.
But for all other cases I don't have an specific publisher like telemetry that is always published to WBA-MONITOR. I can have 50 microservices, so the namespace property is the destination property to publish data dynamically.

**Comment by gustavomassa** [comment]
@pelzerim Thank you for pointing that,

I've thought about it... it makes sense to have a publisher/requester cache, but I think cote could implement that cache logic internally to avoid that kind of issues.

Since my key property is always the hostname value, I just have to cache the namespace property:
```
    private publisherFactory(namespace: string): Publisher {
        try {
            if (namespace === 'WBA-MONITOR') return this.telemetrySender;
            else if (namespace === 'broadcast') return this.broadcastSender;
            else {
                const publisherIndex = this.publisherList.findIndex(x => x.namespace === namespace);
                if (publisherIndex !== -1) return this.publisherList[publisherIndex];
                else {
                    let newPublisher: ServicePublisher = new this.cote.Publisher({ name: `${WBA.config().name} Publisher to ${namespace}`, namespace: namespace, key: WBA.config().hostname }, this.discoveryOptions);
                    newPublisher.namespace = namespace;
                    const newLength = this.publisherList.push(newPublisher);
                    return this.publisherList[newLength - 1];
                }
            }
            return null;

        } catch (ex) {
            WBA.logError(new Error(ex));
            return null;
        }
    }

    private requesterFactory(namespace: string): Requester {
        try {
            const requesterIndex = this.requesterList.findIndex(x => x.namespace === namespace);
            if (requesterIndex !== -1) return this.requesterList[requesterIndex];
            else {
                let newRequester: ServiceRequester = new this.cote.Requester({ name: `${WBA.config().name} Requester to ${namespace}`, namespace: namespace, key: WBA.config().hostname }, this.discoveryOptions);
                newRequester.namespace = namespace;
                const newLength = this.requesterList.push(newRequester);
                return this.requesterList[newLength - 1];
            }
            return null;

        } catch (ex) {
            WBA.logError(new Error(ex));
            return null;
        }
    }
```

> Also do I understand correctly that the publishMsg is responding to a single service with a single request? In that case you should be using Requester/Responders.
Yes and no, publishMsg publishes a message to a single service, using a single publisher without waiting for a response or receiving any message delivered flag from the destination service.

The sendMsg is the one that sends a request to a single service and waits for a response based on a timeout.

**Comment by knoxcard** [comment]
If you are using middleware, just create the service connection once...

app.js
```
  app.cote = require('cote')
  app.geo_client = new app.cote.Requester({
    name: 'Client'
  })
  app.use(require('./middleware')(app))
```

middleware.js
```



```
